core/common/src/main/java/alluxio/underfs/UnderFileSystem.java

```diff
@@ -475,34 +296,7 @@ public OutputStream create(String path, CreateOptions options) throws IOExceptio
    *         empty. Returns {@code null} if this abstract pathname does not denote a directory.
    * @throws IOException if a non-Alluxio error occurs
    */
-  public String[] listRecursive(String path) throws IOException {
-    // Clean the path by creating a URI and turning it back to a string
-    AlluxioURI uri = new AlluxioURI(path);
-    path = uri.toString();
-    List<String> returnPaths = new ArrayList<>();
-    Queue<String> pathsToProcess = new ArrayDeque<>();
-    // We call list initially, so we can return null if the path doesn't denote a directory
-    String[] subpaths = list(path);
-    if (subpaths == null) {
-      return null;
-    } else {
-      for (String subp : subpaths) {
-        pathsToProcess.add(PathUtils.concatPath(path, subp));
-      }
-    }
-    while (!pathsToProcess.isEmpty()) {
-      String p = pathsToProcess.remove();
-      returnPaths.add(p.substring(path.length() + 1));
-      // Add all of its subpaths
-      subpaths = list(p);
-      if (subpaths != null) {
-        for (String subp : subpaths) {
-          pathsToProcess.add(PathUtils.concatPath(p, subp));
-        }
-      }
-    }
-    return returnPaths.toArray(new String[returnPaths.size()]);
-  }
+  String[] listRecursive(String path) throws IOException;
```

Comment:
instead of introducing a method for recursive listing, introduce an options object